### PR TITLE
Storage blob aggregation

### PR DESCRIPTION
**Pull request by vladsud**
**Problem statement:**
1. We generate a ton of (small) blobs, specifically for table scenarios
2. This results in inefficiencies in storage, as well as in snapshot format.
3. This change attempts to remedy most of these problems by packing small blobs into bigger blobs (per data store), as initial step on this path.

**Some key observations:**
1. This change results in bigger summaries (reduced blob de-duping), but substantial less number of nodes and edges in the graph, improving storage perf. This also has indirect impact on snapshot download, which will
   - not be that impacted in terms of bytes over wire due to compression helping with de-duping
   - actually likely reducing current snapshot sizes as metadata is the biggest contributor to size, and amount of metadata is reduced substantially
   - increasing memory foot print, both due to duplication of blobs, as well as creating extra caching layer.
2. 2K cut-off is based on both assumption of granularity of storage ("cluster" size) as well as looking at typical files.
   - This will reduce number of blobs from ~4K to about 600 for Chapter 2 document. Edges will be reduced by a factor of 3x-4x or so.
3. Integrating it at runtime layer for easier supportability.
4. Some of the ugliness in integration points is due to how we use storage and snapshots, conflating concerns, casting incorrectly types, etc.
5. While working on it, I keep finding bugs in existing code. For example you see work around for bufferToString() - I'll follow up on that separately.
6. Packing algorithm and its limitations can be improved incrementally in future, allowing more incremental summaries (removing the need for all DDSs or nothing to be summarized by data store).
    - the key here is that unpacking logic is rather generic and should not change, thus allowing this code to be future proofed.

**What's left:**
1. UTs
2. For some reason I can't hit handle reuse case in my testing, so not sure if  my handling of SummaryType.Handle is correct - need to figure out why I'm not hitting it (i.e. summaries are all full).

What to pay attention to:
1. Integration and unpacking code - it needs to go in first (keeping packing code not enabled) and get new bits in production to saturation, before we can enable packing itself.

**Future improvements:**
1. Remove limitations (more incremental summaries)
2. Consider moving aggregated blobs one level up, reducing depth of tree and thus cutting number of edges for Chapter2 document by another factor of 2x.

### CODE REVIEW

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 1]
```diff
@@ -481,13 +482,29 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
 
+        let storage = context.storage;
+        if (context.baseSnapshot) {
+            // This will patch snapshot in place!
+            // If storage is provided, it will wrap storage with BlobAggregatorStorage that can
+            // pack & unpack aggregated blobs.
+            // Note that if storage is provided later by loader layer, we will wrap storage in this.storage getter.
+            // BlobAggregatorStorage is smart enough for double-wrapping to be no-op
+            if (context.storage) {
+                const aggrStorage = BlobAggregatorStorage.wrap(context.storage);
```
Comment (curtisman) [inline packages/runtime/container-runtime/src/containerRuntime.ts:493]: We should add a way for the storage to declare whether we want this behavior or not.  This behavior might not be desirable for r11s.

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+
+        const newSummary: ISummaryTree = {...summary};
+        for (const key of Object.keys(newSummary.tree)) {
+            const obj = newSummary.tree[key];
+            // Get path relative to root of data store (where we do aggregation)
+            const newPath = startingLevel ? key : `${path}/${key}`;
+            switch (obj.type) {
+                case SummaryType.Tree:
+                    // If client created empty tree, keep it as is
+                    // Also do not package search blobs - they are path of storage contract
```
Comment (curtisman) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:304]: path => part (sp)

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        tree.blobs[id] = fromUtf8ToBase64(content);
+    }
+}
+
+/*
+ * Snapshot packer and extractor.
+ * When summary is written it will find and aggregate small blobs into bigger blobs
+ * When snapshot is read, it will unpack aggregated blobs and provide them transparently to caller.
+ */
+export class BlobAggregatorStorage extends SnapshotExtractor implements IDocumentStorageService {
```
Comment (curtisman) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:165]: fluid-fetch integration will be necessary.

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+}
+
+ /**
+  * Class responsible for aggregating smaller blobs into one and unpacking it later on.
+  */
+class BlobAggregator {
+    private readonly content: [string, string][]= [];
+
+    public addBlob(key: string, content: string) {
+        this.content.push([key, content]);
```
Comment (curtisman) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:60]: Should we try to dedup contents?  (Small blobs are more likely to be dedup-able too)

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts (hunk 3 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobAggregationStorage.ts:165]: It's good point, though I wonder - where does it stop? I.e. this "interpretation" layer that is on top of storage.
fluid-fetch tool already cracks some things, so probably it's fine, but it feels like it should be separate tool to give storage representation some "meaning".
Not pushing back, just saying that it adds some complexity and mixing concerns.

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts (hunk 4 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobAggregationStorage.ts:60]: We could, but note that aggregation (as it's built right now) works on data store level.
So the only real dedupping we can get is 2 attributes for a map DDS when data store has 2 maps ("average" payload has Scriptors that use one directory, 2 maps, and 2 Sequence).

I'd prefer to go for simplicity (given very low gain), but I'll also contradict myself and say that if we believe we need it, we better of building it into unpacking algorithm sooner than later, to have faster deployment timing in future.

Note that bytes over the wire are not affected that much thanks to compression. And in-memory saving is low.
But as a general feature - sure, it sounds great to have it!

I think to punt for now, but open (post-GA) item to do de-duping.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 5]
```diff
@@ -657,7 +657,13 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
                 // (i.e. it has a base mapping) - then we go ahead and summarize
                 return isAttached;
             }).map(async ([contextId, context]) => {
-                const contextSummary = await context.summarize(fullTree, trackState);
+                // If BlobAggregatorStorage is engaged, we have to write full summary for data stores
+                // BlobAggregatorStorage relies on this behavior, as it aggregates blobs across DDSs.
+                // Not generating full summary will mean data loss, as we will overwrite aggregate blob in new summary,
+                // and any virtual blobs that stayed (for unchanged DDSs) will need aggregate blob in previous summary
+                // that is no longer present in this summary.
+                // This is temporal limitation that cna be lifted in future once BlobAggregatorStorage becomes smarter.
```
Comment (jatgarg) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:665]: >cna [](start = 52, length = 3)

cna -> "can"

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+                    const pathSplit = path.split("/");
+                    let subTree = snapshot;
+                    for (const subPath of pathSplit.slice(0, pathSplit.length - 1)) {
+                        if (subTree.trees[subPath] === undefined) {
+                            subTree.trees[subPath] = { blobs: {}, commits: {}, trees: {}};
+                        }
+                        subTree = subTree.trees[subPath];
+                    }
+                    const blobName = pathSplit[pathSplit.length - 1];
+                    assert(subTree.blobs[blobName] === undefined);
```
Comment (jatgarg) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:124]: >assert [](start = 20, length = 6)

Please add a message for this assert as it is easier to find problems later.

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+ */
+class SnapshotExtractorInPlace extends SnapshotExtractor {
+    public async getBlob(id: string, tree: ISnapshotTree): Promise<ArrayBufferLike> {
+        const blob = tree.blobs[id];
+        assert(blob !== undefined);
+        return stringToBuffer(blob, "base64");
+    }
+
+    public setBlob(id: string, tree: ISnapshotTree, content: string) {
+        assert(tree.blobs[id] === undefined);
```
Comment (jatgarg) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:155]: Same here, please add an assert message.

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+ * 2. map blob ID to blob content
+ * #2 is what storage (IDocumentStorageService) is for, but in places where we do not have it
+ * (like loading serialized earlier draft content), blob content is put directly into snapshot.
+ * Ideally this should be fixed by using BlobCacheStorageService or something similar and
+ * fixing existing flows to allow switching of storage.
+ */
+class SnapshotExtractorInPlace extends SnapshotExtractor {
+    public async getBlob(id: string, tree: ISnapshotTree): Promise<ArrayBufferLike> {
+        const blob = tree.blobs[id];
+        assert(blob !== undefined);
```
Comment (jatgarg) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:150]: >s [](start = 9, length = 1)

Same here, please add an assert message.

**Comment by vladsud** [comment]
Other than adding UT, it's ready to go. Did a bunch of testing - looks pretty good :) Would appreciate people taking a look.
Q: should implementation stay in container-runtime or move to driver-utils?

**Comment by jatgarg** [comment]
> Other than adding UT, it's ready to go. Did a bunch of testing - looks pretty good :) Would appreciate people taking a look.
> Q: should implementation stay in container-runtime or move to driver-utils?

I think we should keep it in runtime. Because:
1.) It is happening at runtime layer or used by runtime layer to interact with snapshot while packing or unpacking. Drivers does not know anything about it. So it doesn't makes sense for me to keep it in driver-utils.

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts (hunk 3 above)
Comment (curtisman) [reply packages/runtime/container-runtime/src/blobAggregationStorage.ts:164]: That's very true, although fluid-fetch probably should able to report on various layer interpretation.

**Comment by curtisman** [comment]
I think that logically this should be in the driver layer.   I believe current;y this "bundled" directly with the runtime only because we worry about upgrade velocity, but it should be the driver's decision to include this "optimization" or not based on the server characteristic.

**Comment by tanviraumi** [comment]
I agree that it should be in the driver layer. It really depends on the specific storage provider and whether it wants to have this optimization or not.

**Comment by agarwal-navin** [comment]
+1 to this being part of driver. This optimization is specific to SPO today and while others may want it, it should be optional.

File: packages/loader/container-loader/src/container.ts [hunk 9]
```diff
@@ -894,8 +893,11 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
         }
 
         if (blobs.size > 0) {
-            this.blobsCacheStorageService =
+            const blobSize = this.storageService.policies?.minBlobSize;
+            this._storageService =
                 new BlobCacheStorageService(this.storageService, blobs);
+            // ensure we did not lose that policy in the process of wrapping
+            assert(blobSize === this._storageService.policies?.minBlobSize);
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/container.ts:900]: assert message

File: packages/loader/driver-definitions/src/storage.ts [hunk 10]
```diff
@@ -208,19 +221,10 @@ export enum LoaderCachingPolicy {
     Prefetch,
 }
 
-export interface IDocumentServicePolicies {
```
Comment (anthony-murphy) [inline packages/loader/driver-definitions/src/storage.ts:211]: the current plan for storage only mode uses this. i'm fine moving caching, but think we should keep this as well

File: packages/loader/container-loader/src/containerContext.ts [hunk 11]
```diff
@@ -145,7 +145,7 @@ export class ContainerContext implements IContainerContext {
         return this._baseSnapshot;
     }
 
-    public get storage(): IDocumentStorageService | undefined | null {
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/containerContext.ts:148]: >| [](start = 62, length = 1)

why do we need to add a null state here?

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 12]
```diff
@@ -0,0 +1,388 @@
+/*!
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:1]: can you add unit tests for this file

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+
+// Gate that when flipped, instructs to compress small blobs.
+// We have to first ship with this gate off, such that we can get to saturation bits
+// that can understand compressed format. And only after that flip it.
+function gatesAllowPacking() {
+    // Leave override for testing purposes
+    if (typeof localStorage === "object" && localStorage !== null && localStorage.FluidAggregateBlobs === "1") {
+        return true;
+    }
```
Comment (anthony-murphy) [inline packages/runtime/container-runtime/src/blobAggregationStorage.ts:35]: i don't like this here. can we use a loader option, we can then se the loader option based on localStorage.

File: packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts [hunk 14]
```diff
@@ -201,6 +202,23 @@ class BlobCache {
...
+        // Runtime creating many small blobs results in sub-optimal perf.
+        // 2K seems like the sweat spot:
+        // The smaller the number, less blobs we aggregate. Most storages are very likely to have notion
+        // of minimal "cluster" size, so having small blobs is wasteful
+        // At the same time increasing the limit ensure that more blobs with user content are aggregated,
+        // reducing possibility for de-duping of same blobs (i.e. .attributes rolled into aggregate blob
+        // are not reused across data stores, or even within data store, resulting in duplication of content)
+        // Note that duplication of content should not have significant impact for bytes over wire as
+        // compression of http payload mostly takes care of it, but it does impact storage size and in-memory sizes.
+        minBlobSize: 2048,
```
Comment (anthony-murphy) [inline packages/drivers/odsp-driver/src/odspDocumentStorageManager.ts:219]: >minBlobSize [](start = 8, length = 11)

might also want to turn this on for the LocalDocumentStorageService as that will give a lot of test coverage

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 1 above)
Comment (anthony-murphy) [reply packages/runtime/container-runtime/src/containerRuntime.ts]: i think you always need it. it doesn't write anything compressed if minBlobSize in undefined, but regardless can read the compressed format

---
In reply to: [569728765](https://github.com/microsoft/FluidFramework/pull/5032#discussion_r569728765) [](ancestors = 569728765)

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts (hunk 13 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobAggregationStorage.ts]: This is strictly for testing purposes. I.e. be able to overwrite setting locally in dev tools.
It is also temporarily here, for the moment when client bits containing ability to read aggregated blobs are propagating, and before we pull the plug to start writing. Once writing is enabled in prod, that code would be deleted.
I'm not sure I see a good reason to expose it on loader / make code changes in consumers. If anything, it should be an option in SPO driver factory, not loader.

File: packages/loader/container-loader/src/containerContext.ts (hunk 11 above)
Comment (vladsud) [reply packages/loader/container-loader/src/containerContext.ts:148]: I have no clue why null was here, it was not used, so I deleted it (your comment suggests that I'm adding it, which is not the case)

File: packages/loader/driver-definitions/src/storage.ts (hunk 10 above)
Comment (vladsud) [reply packages/loader/driver-definitions/src/storage.ts:211]: If we need policies on service as a whole, we can always add it in future. For not, both policies are for storage part of service only, so it makes sense to move it.

File: packages/loader/container-loader/src/deltaManager.ts [hunk 15]
```diff
@@ -315,11 +315,15 @@ export class DeltaManager
...
         // Enable prefetching for the service unless it has a caching policy set otherwise:
-        if (service.policies?.caching !== LoaderCachingPolicy.NoCaching) {
+        if (storageService.policies?.caching !== LoaderCachingPolicy.NoCaching) {
             storageService = new PrefetchDocumentStorageService(storageService);
         }
 
         this.storageService = new RetriableDocumentStorageService(storageService, this, this.logger);
+
+        // ensure we did not lose that policy in the process of wrapping
+        assert(storageService.policies?.minBlobSize === this.storageService.policies?.minBlobSize);
```
Comment (anthony-murphy) [inline packages/loader/container-loader/src/deltaManager.ts:325]: message

**Comment by anthony-murphy** [comment]
because the driver is part of the loader which is statically linked into things like native and mobile apps we have very little agility in terms of bug fixing. given that i'd very much prefer this in the contianerRuntimer layer

**Comment by anthony-murphy** [comment]
would it be possible to turn this on at some level in our snapshot replay tests so we have a locked version of the format persisted for testing?

**Review by anthony-murphy — APPROVED** [review]

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 16]
```diff
@@ -697,7 +697,13 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
                 // (i.e. it has a base mapping) - then we go ahead and summarize
                 return isAttached;
             }).map(async ([contextId, context]) => {
-                const contextSummary = await context.summarize(fullTree, trackState);
+                // If BlobAggregationStorage is engaged, we have to write full summary for data stores
```
Comment (vladsud) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:700]: FYI - Opened https://github.com/microsoft/FluidFramework/issues/5165 to track this issue of incremental snapshots

**Comment by curtisman** [comment]
> because the driver is part of the loader which is statically linked into things like native and mobile apps we have very little agility in terms of bug fixing. given that I'd very much prefer this in the contianerRuntimer layer. i don't really care where the pacakges lives, as long as the code is run in the containerRuntime and not the loader or driver

I understand the concern about the speed of updating the loader/driver in our scenarios. However, putting in the runtime will create a bigger problem long term as the number of container developer glows.   The number of container solution would far exceed the number of host and most of them will be out of our control. Tracking and asking all of these partners to update will be a tall order and will not be a speedy process either.

Logically, I also think that it is driver specific, as the optimization is very storage dependent.  I am ok if this is temporary bundled with the runtime, but I'd' much rather to improve the speed of loader/drvier update if possible rather then creating a problem that is harder to manage longer term.

**Comment by vladsud** [comment]
I've opened https://github.com/microsoft/FluidFramework/issues/5166 to move this layer into drivers (assigned it for now to May sprint). I agree that it belongs there, but I think the key pre-condition - we need to have a solution for how we update drivers, designed and implemented.

I think the key here - build things in layers. And leave linking (aka packing) decision to a layer time (and make it as much as possible to be independent from feature implementation).
I.e. till now, most of the code in ODSP driver was built as add-on, i.e. integrated into the core of the driver, making it really hard to enable/disable pieces or substitute them. Building this logic as an adapter on the storage allows us easily add it to any storage, thus making decisions about where it belongs easy to flip.

I believe other things (like blob de-dupping) can be redesigned to be the same (i.e. adapters), and we may chose to ship them temporarily at runtime layer (or maybe not).

File: packages/runtime/container-runtime/src/blobAggregationStorage.ts (hunk 3 above)
Comment (vladsud) [reply packages/runtime/container-runtime/src/blobAggregationStorage.ts]: Opened https://github.com/microsoft/FluidFramework/issues/5167 to do fast follow up - it needs to be done before I enable write part of this change (which should happen in ~2 releases)

**Pull request merged**